[RegAlloc] Scale the spill weight by target factor

wangpc-pp opened the pull request:
Currently, the spill weight is only determined by isDef/isUse and
block frequency. However, for registers with different register
classes, the costs of spilling them are different.

For example, for `LMUL>1` registers (in which, several physical
registers compound a bigger logical register), the costs are larger
than `LMUL=1` case (in which, there is only one physical register).

To solve this problem, a new target hook `getSpillWeightScaleFactor`
is added. Targets can override the default factor (which is `1.0`)
according to the register class.

For RISC-V, the factors are set to the `RegClassWeight` which is
used to track register pressure. The values of `RegClassWeight`
happen to be the number of register units.

I believe all of the targets with compounded registers can benefit
from this change, but only RISC-V is customized in this patch since
it has widely been agreed to do so. The other targets need more
performance data to go further.

Partially fixes #113489.

wangpc-pp commented on llvm/test/CodeGen/RISCV/rvv/bitreverse-vp.ll:
```diff
@@ -2307,36 +2307,38 @@ define <vscale x 7 x i64> @vp_bitreverse_nxv7i64(<vscale x 7 x i64> %va, <vscale
+; RV32-NEXT:    vand.vv v16, v8, v16, v0.t
+; RV32-NEXT:    vsll.vi v16, v16, 8, v0.t
+; RV32-NEXT:    vor.vv v16, v24, v16, v0.t
+; RV32-NEXT:    csrr a5, vlenb
```
This can be considered as a small regression because we have more reads of `vscale` (vlenb).

lukel97 reviewed (commented):
Thanks for looking at this, I also tried this a few months ago, and saw the same test diff where we're able to remove a lot of LMUL 8 spills. 

However whilst it helps the in-tree tests a lot when I tried it out on SPEC CPU 2017 I got a 50% increase in spills generated overall, have you been able to check if you run into this too?

My approach also didn't use the reg class weight but instead I scaled the weight by the physical register size in bytes. From what I remember investigating the register class weight was related to register pressure calculation, which seemed orthogonal to spilling

wangpc-pp commented:
> Thanks for looking at this, I also tried this a few months ago, and saw the same test diff where we're able to remove a lot of LMUL 8 spills.
> 
> However whilst it helps the in-tree tests a lot when I tried it out on SPEC CPU 2017 I got a 50% increase in spills generated overall, have you been able to check if you run into this too?

Here is the result of statistics (`-march=rv64gcv`):
```
# utils/compare.py -m regalloc.NumSpills baseline.json experimental.json                            
Tests: 10                                                                                                                                                                                                    
Metric: regalloc.NumSpills                                                                                                                                                                                   
                                                                                                                                                                                                             
Program                                       regalloc.NumSpills                                                                                                                                             
                                              baseline           experimental diff                                                                                                                           
500.perlbench_r/500.perlbench_r.test           4255.00            4255.00      0.0%                                                                                                                          
505.mcf_r/505.mcf_r.test                        128.00             128.00      0.0%                                                                                                                          
520.omnetpp_r/520.omnetpp_r.test                971.00             971.00      0.0%                                                                                                                          
523.xalancbmk_r/523.xalancbmk_r.test           1578.00            1578.00      0.0%                                                                                                                          
525.x264_r/525.x264_r.test                     2027.00            2027.00      0.0%                                                                                                                          
531.deepsjeng_r/531.deepsjeng_r.test            355.00             355.00      0.0%                                                                                                                          
541.leela_r/541.leela_r.test                    361.00             361.00      0.0%                                                                                                                          
557.xz_r/557.xz_r.test                          296.00             296.00      0.0%                                                                                                                          
502.gcc_r/502.gcc_r.test                      13739.00           13735.00     -0.0%                                                                                                                          
999.specrand_ir/999.specrand_ir.test              0.00               0.00                                                                                                                                    
                           Geomean difference                                 -0.0%                                                                                                                          
      regalloc.NumSpills                                                                                                                                                                                     
run             baseline  experimental      diff                                                                                                                                                             
count  10.000000          10.000000     9.000000                                                                                                                                                             
mean   2371.000000        2370.600000  -0.000032                                                                                                                                                             
std    4197.231045        4196.027306   0.000097                                                                                                                                                             
min    0.000000           0.000000     -0.000291                                                                                                                                                             
25%    310.750000         310.750000    0.000000                                                                                                                                                             
50%    666.000000         666.000000    0.000000
75%    1914.750000        1914.750000   0.000000
max    13739.000000       13735.000000  0.000000

# utils/compare.py -m regalloc.NumReloads baseline.json experimental.json 
Tests: 10
Metric: regalloc.NumReloads

Program                                       regalloc.NumReloads                   
                                              baseline            experimental diff 
500.perlbench_r/500.perlbench_r.test           9790.00             9790.00      0.0%
505.mcf_r/505.mcf_r.test                        438.00              438.00      0.0%
520.omnetpp_r/520.omnetpp_r.test               1474.00             1474.00      0.0%
523.xalancbmk_r/523.xalancbmk_r.test           2511.00             2511.00      0.0%
525.x264_r/525.x264_r.test                     4694.00             4694.00      0.0%
531.deepsjeng_r/531.deepsjeng_r.test            674.00              674.00      0.0%
541.leela_r/541.leela_r.test                    509.00              509.00      0.0%
557.xz_r/557.xz_r.test                          631.00              631.00      0.0%
502.gcc_r/502.gcc_r.test                      31924.00            31920.00     -0.0%
999.specrand_ir/999.specrand_ir.test              0.00                0.00          
                           Geomean difference                                  -0.0%
      regalloc.NumReloads                        
run              baseline  experimental      diff
count  10.000000           10.000000     9.000000
mean   5264.500000         5264.100000  -0.000014
std    9829.948957         9828.743601   0.000042
min    0.000000            0.000000     -0.000125
25%    539.500000          539.500000    0.000000
50%    1074.000000         1074.000000   0.000000
75%    4148.250000         4148.250000   0.000000
max    31924.000000        31920.000000  0.000000
```

I don't see the increase of spills, but the decrease is also imperceptible.

> My approach also didn't use the reg class weight but instead I scaled the weight by the physical register size in bytes. From what I remember investigating the register class weight was related to register pressure calculation, which seemed orthogonal to spilling

Scaling the weight by bytes may be over estimated. I wanted to scale the weight by the number of physical registers, which is accidentally the register class weight.

lukel97 commented:
My bad, I was getting two things conflated. I had also tried a different approach where instead of adjusting the spill weights, I changed the eviction advisor to aggressively perform local reassignment (i.e. when trying to evict an m8 register, it would now try to evict an m1 register instead if it could be immediately assigned a register): https://github.com/lukel97/llvm-project/commit/4075bfc46cf93978d3ca87442c725e737d973546. Off the top of my head I think this was the one that led to all the spilling in SPEC (but also fixed the spilling in-tree)

But I checked out this PR locally and tried it with -march=rva23u64 -O3 -flto and still got a slight increase in the number of spills. I think with LTO there's more areas of high register pressure that this affects.

```
Program            regalloc.NumSpills                regalloc.NumReloads               
                   lhs                rhs      diff  lhs                 rhs      diff 
   531.deepsjeng_r   370.00             376.00  1.6%   793.00              800.00  0.9%
   631.deepsjeng_s   370.00             376.00  1.6%   793.00              800.00  0.9%
     538.imagick_r  4406.00            4423.00  0.4% 14425.00            14467.00  0.3%
     638.imagick_s  4406.00            4423.00  0.4% 14425.00            14467.00  0.3%
     526.blender_r 21007.00           21034.00  0.1% 47955.00            47987.00  0.1%
         602.gcc_s 19380.00           19385.00  0.0% 52642.00            52658.00  0.0%
         502.gcc_r 19380.00           19385.00  0.0% 52642.00            52658.00  0.0%
        508.namd_r  5440.00            5440.00  0.0% 12717.00            12717.00  0.0%
       641.leela_s   246.00             246.00  0.0%   482.00              482.00  0.0%
   623.xalancbmk_s  2258.00            2258.00  0.0%  4460.00             4460.00  0.0%
     620.omnetpp_s   997.00             997.00  0.0%  2837.00             2837.00  0.0%
         605.mcf_s    58.00              58.00  0.0%   107.00              107.00  0.0%
   600.perlbench_s  5432.00            5432.00  0.0% 14193.00            14193.00  0.0%
          557.xz_r   316.00             316.00  0.0%   637.00              637.00  0.0%
       541.leela_r   246.00             246.00  0.0%   482.00              482.00  0.0%
   523.xalancbmk_r  2258.00            2258.00  0.0%  4460.00             4460.00  0.0%
      510.parest_r  4565.00            4565.00  0.0% 11879.00            11879.00  0.0%
     520.omnetpp_r   997.00             997.00  0.0%  2837.00             2837.00  0.0%
         505.mcf_r    58.00              58.00  0.0%   107.00              107.00  0.0%
   500.perlbench_r  5432.00            5432.00  0.0% 14193.00            14193.00  0.0%
         644.nab_s   354.00             354.00  0.0%   787.00              787.00  0.0%
         619.lbm_s    70.00              70.00  0.0%    89.00               89.00  0.0%
         544.nab_r   354.00             354.00  0.0%   787.00              787.00  0.0%
         519.lbm_r    72.00              72.00  0.0%    91.00               91.00  0.0%
      511.povray_r  2453.00            2453.00  0.0%  5813.00             5813.00  0.0%
          657.xz_s   316.00             316.00  0.0%   637.00              637.00  0.0%
        525.x264_r  2456.00            2454.00 -0.1%  5504.00             5503.00 -0.0%
        625.x264_s  2456.00            2454.00 -0.1%  5504.00             5503.00 -0.0%
   997.specrand_fr     0.00               0.00                                         
   996.specrand_fs     0.00               0.00                                         
   999.specrand_ir     0.00               0.00                                         
   998.specrand_is     0.00               0.00                                         
Geomean difference                              0.1%                               0.1%
```

With that said, I'm not sure if these spills or reloads are signficant for performance. And on the llvm-test-suite loop vectorization micro benchmarks it actually removes all of the spills.

```
                   test-suite :: MicroBenchmarks/LoopVectorization/LoopVectorizationBenchmarks.test   391.00               0.00 -100.0%   905.00                     -100.0%
                    test-suite :: MicroBenchmarks/LoopVectorization/LoopInterleavingBenchmarks.test   391.00               0.00 -100.0%   905.00                     -100.0%
```

I will try and a do a run on deepsjeng and imagick to check this, but I would be overall in favour of this. 

> I wanted to scale the weight by the number of physical registers, which is accidentally the register class weight.

For what it's worth, I also previously tried scaling it by the number of reg units: https://github.com/lukel97/llvm-project/commit/65bd27ec3db908ee24abd40539508875800165df 

But I think this might actually be the same thing that the reg class weight should reflect?

```
  // Weight override for register pressure calculation. This is the value
  // TargetRegisterClass::getRegClassWeight() will return. The weight is in
  // units of pressure for this register class. If unset tablegen will
  // calculate a weight based on a number of register units in this register
  // class registers. The weight is per register.
  int Weight = ?;
```

preames commented:
I am (conceptually) in favor of this direction.  I haven't looked at the code yet in detail, and plan on deferring that until the empirical measurement discussion has converged.

preames commented:
@lukel97 Any update on the perf measurements you mentioned in your last comment?

(Also FYI @mikhailramalho)

bzEq commented:
Have you tried taking LaneBitMask into account rather than adding the `Factor`? The LaneBitMask indicates concrete regunits occupied by the RegClass.

bzEq commented:
> But I think this might actually be the same thing that the reg class weight should reflect?

  // Weight override for register pressure calculation. This is the value
  // TargetRegisterClass::getRegClassWeight() will return. The weight is in
  // units of pressure for this register class. If unset tablegen will
  // calculate a weight based on a number of register units in this register
  // class registers. The weight is per register.
  int Weight = ?;

I think it's viable. Except that regunit introduced by aliases might not be included in calculating the SpillWeight.

wangpc-pp commented:
> Have you tried taking LaneBitMask into account? The LaneBitMask indicates regunits occupied by the RegClass.

Yes, I tried it before. It is just the same as https://github.com/lukel97/llvm-project/commit/65bd27ec3db908ee24abd40539508875800165df I think.

bzEq commented on llvm/include/llvm/CodeGen/TargetRegisterInfo.h:
```diff
@@ -926,6 +926,9 @@ class TargetRegisterInfo : public MCRegisterInfo {
   virtual const int *getRegUnitPressureSets(unsigned RegUnit) const = 0;
 
+  /// Get the factor of spill weight for this register class.
+  virtual unsigned getSpillWeightFactor(const TargetRegisterClass *RC) const;
```
nit
```suggestion
  virtual unsigned getSpillWeightScaleFactor(const TargetRegisterClass *RC) const;
```

preames commented:
ping @lukel97 and @mikhailramalho re: BP3 perf measurements.

lukel97 commented:
> ping @lukel97 and @mikhailramalho re: BP3 perf measurements.

I've kicked off a run now, will report back when the results are ready. 

I'm wondering if this is the right place to discourage the spilling high LMUL registers. There's multiple ways we could do this and increasing the spill weights is just one, e.g. [Local reassignment](https://github.com/lukel97/llvm-project/commit/4075bfc46cf93978d3ca87442c725e737d973546) is another. I have a feeling that the spill weights might really be intended to calculate the frequency of spills, not the overall cost. E.g. there is the possibility that spilling one M8 register allows 8 M1 registers to avoid spills.

FWIW this is approach is also what I originally tried but I think it would be good to hear some thoughts from people that have worked more with the greedy register allocator.

arsenm commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -803,6 +803,11 @@ RISCVRegisterInfo::getRegisterCostTableIndex(const MachineFunction &MF) const {
 
+unsigned
+RISCVRegisterInfo::getSpillWeightFactor(const TargetRegisterClass *RC) const {
+  return getRegClassWeight(RC).RegWeight;
```
Why does this need to be a virtual function if it's just reading a property out of the TargetRegisterClass

wangpc-pp replied:
Because I don't know if all targets should use this, so I made it a target hook and overrided it for RISC-V.

arsenm replied:
Do not do this. Just use the field

wangpc-pp replied:
You mean I should apply this change to all the targets (the shared part)?

arsenm replied:
Yes. The opt-out is just use a factor of 1

wangpc-pp replied:
The change will be huge if this is applied to all targets, and targets may have different factors. So I think we can keep the hook and let the targe make the decision.

arsenm replied:
No, do not increase the target API surface out of fear of changing other targets. This is a large problem for llvm where everyone just adds some new single purpose knob that nothing else will ever implement. There must be attempts at universal infrastructure improvements. The targets may have different weights, but there is already the mechanism here to control it. Just use it instead of adding another undiscoverable, redundant override to implement

wangpc-pp replied:
Done.

preames commented on llvm/test/CodeGen/RISCV/rvv/ctpop-vp.ll:
```diff
@@ -2038,231 +2037,145 @@ define <vscale x 16 x i64> @vp_ctpop_nxv16i64(<vscale x 16 x i64> %va, <vscale x
 ; RV32-NEXT:    add a3, sp, a3
 ; RV32-NEXT:    addi a3, a3, 16
-; RV32-NEXT:    vs8r.v v16, (a3) # Unknown-size Folded Spill
+; RV32-NEXT:    vs8r.v v8, (a3) # Unknown-size Folded Spill
```
This case looks suspicious.  We have existing logic to rematerialize a vmv.v.x instead of spilling it.  Why didn't that kick in here?  Both before and after.

preames commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-floor-vp.ll:
```diff
@@ -763,11 +757,7 @@ define <32 x double> @vp_floor_v32f64(<32 x double> %va, <32 x i1> %m, i32 zeroe
-; CHECK-NEXT:    addi a2, sp, 16
-; CHECK-NEXT:    vs8r.v v24, (a2) # Unknown-size Folded Spill
 ; CHECK-NEXT:    fsrm a1
-; CHECK-NEXT:    addi a1, sp, 16
```
This is weird?  Is fsrm maybe marked as invalidating vector registers?  We have a spill immediately before a fill from the same address a few lines later?

preames reviewed (commented):
A few off-topic observations from the diffs.

lukel97 commented:
I'm not seeing any measurable runtime difference on SPEC CPU 2017 or MicroBenchmarks/LoopVectorization

qcolombet reviewed (commented):
The direction sounds fine to me.

wangpc-pp commented:
> E.g. there is the possibility that spilling one M8 register allows 8 M1 registers to avoid spills.

I don't think this is possible in current greedy RA (please correct me if my understanding is wrong). When spilling, we compare the weights of candidates one by one, because we can easily compare the live intervals of two registers. But when it comes to 8 registers, there are some problems: how can we compare the live intervals? how can compare the weight of one register with the weight of 8 registers (that means, how can we calculate one weight from 8 weights?), etc.

wangpc-pp commented:
I believe this is the right direction (and I think most people have echoed the same thought). Now this change is applied to the infrastructure code, which means all the targets will be influenced. I have a rough scan of the diffs, I think most of them are positive changes, but I need more eyes on other targets besides RISC-V.

lukel97 commented:
> > E.g. there is the possibility that spilling one M8 register allows 8 M1 registers to avoid spills.
> 
> I don't think this is possible in current greedy RA (please correct me if my understanding is wrong). When spilling, we compare the weights of candidates one by one, because we can easily compare the live intervals of two registers. But when it comes to 8 registers, there are some problems: how can we compare the live intervals? how can compare the weight of one register with the weight of 8 registers (that means, how can we calculate one weight from 8 weights?), etc.

I agree that the eviction advisor isn't currently taking regunits/register size into account, but I don't see how that wouldn't make it possible for an LMUL 8 spill to avoid 8 M1 spills.

From my understanding of `DefaultEvictionAdvisor::canEvictInterferenceBasedOnCost`, when choosing a register to evict it just goes through the registers in the allocation order and picks the one with the smallest spill weight in its interferences.

So if an interfering LMUL 8 LiveInterval has a smaller spill weight, e.g. is only used once, it may get evicted instead of an LMUL 1 LiveInterval with say two uses (higher spill weight). 

If the LMUL 8 LiveInterval is evicted and subsequently spilled, then 8 more regunits become free which can prevent further spilling.

So the eviction advisor doesn't take into account the number of regunits (size of the registers), only the spill weights (the frequency with which they're used).

I think the bit I would like to get clarity on is, are we conflating two separate metrics here, the frequency with which a LiveInterval would need to be spilled and reloaded, vs the size of the spills? Because currently spill weights are based on the former, and this PR is now adding the latter

wangpc-pp commented:
> > > E.g. there is the possibility that spilling one M8 register allows 8 M1 registers to avoid spills.
> > 
> > 
> > I don't think this is possible in current greedy RA (please correct me if my understanding is wrong). When spilling, we compare the weights of candidates one by one, because we can easily compare the live intervals of two registers. But when it comes to 8 registers, there are some problems: how can we compare the live intervals? how can compare the weight of one register with the weight of 8 registers (that means, how can we calculate one weight from 8 weights?), etc.
> 
> I agree that the eviction advisor isn't currently taking regunits/register size into account, but I don't see how that wouldn't make it possible for an LMUL 8 spill to avoid 8 M1 spills.
> 
> From my understanding of `DefaultEvictionAdvisor::canEvictInterferenceBasedOnCost`, when choosing a register to evict it just goes through the registers in the allocation order and picks the one with the smallest spill weight in its interferences.
> 
> So if an interfering LMUL 8 LiveInterval has a smaller spill weight, e.g. is only used once, it may get evicted instead of an LMUL 1 LiveInterval with say two uses (higher spill weight).
> 
> If the LMUL 8 LiveInterval is evicted and subsequently spilled, then 8 more regunits become free which can prevent further spilling.
> 
> So the eviction advisor doesn't take into account the number of regunits (size of the registers), only the spill weights (the frequency with which they're used).
> 
> I think the bit I would like to get clarity on is, are we conflating two separate metrics here, the frequency with which a LiveInterval would need to be spilled and reloaded, vs the size of the spills? Because currently spill weights are based on the former, and this PR is now adding the latter

Oh I know your point here! Here is my thinking: I think the `frequency` and the `size` here are the same thing as long as we are using reg units to scale the weight. For example, for a LMUL 8 register, we can consider it as that all its 8 registers have one use, `frequency+=8` and also `size+=8`. Maybe thinking in this way will make this patch more straightforward?

uweigand commented:
@JonPsson1 the cond-move-regalloc-hints.mir change here looks like a regression.  Can you have a look why this happens?

uweigand commented:
> I have some impressions about this test `cond-move-regalloc-hints.mir` when fixing these tests that are not using `utils/update_llc_test_checks.py`. :-)

Ah, you're right, that is an improvement.  The test probably should switch to auto-generated, so that it's clearer what it tests ...

arsenm commented on llvm/test/CodeGen/AMDGPU/load-global-i16.ll:
```diff
@@ -7357,6 +7357,20 @@ define amdgpu_kernel void @global_zextload_v32i16_to_v32i64(ptr addrspace(1) %ou
 ; GCN-NOHSA-SI-NEXT:    buffer_load_dword v15, off, s[12:15], 0 offset:12 ; 4-byte Folded Reload
 ; GCN-NOHSA-SI-NEXT:    s_waitcnt vmcnt(2)
 ; GCN-NOHSA-SI-NEXT:    v_mov_b32_e32 v13, v39
+; GCN-NOHSA-SI-NEXT:    buffer_store_dword v12, off, s[12:15], 0 ; 4-byte Folded Spill
```
This file is pure regression. What happened here?

wangpc-pp replied:
I know nothing about AMDGPU target, can someone take a look at this?

arsenm commented on llvm/test/CodeGen/AMDGPU/remat-smrd.mir:
```diff
@@ -295,13 +295,21 @@ body:             |
-    ; GCN-NEXT: renamable $sgpr1 = S_LOAD_DWORD_SGPR renamable $sgpr2_sgpr3, renamable $sgpr0, 0 :: (dereferenceable invariant load (s32), addrspace 4)
-    ; GCN-NEXT: S_NOP 0, implicit killed renamable $sgpr1
-    ; GCN-NEXT: renamable $sgpr1 = S_LOAD_DWORD_SGPR renamable $sgpr2_sgpr3, renamable $sgpr0, 0 :: (dereferenceable invariant load (s32), addrspace 4)
+    ; GCN-NEXT: SI_SPILL_S32_SAVE $sgpr10, %stack.0, implicit $exec, implicit $sp_reg :: (store (s32) into %stack.0, addrspace 5)
```
Regression. This looks like rematerialization didn't happen?

arsenm replied:
This test has been defeated. Rematerialization looks like it isn't happen at all. I think this needs to be fixi

arsenm commented on llvm/test/CodeGen/AMDGPU/scc-clobbered-sgpr-to-vmem-spill.ll:
```diff
@@ -83,218 +83,220 @@ define amdgpu_kernel void @kernel0(ptr addrspace(1) %out, i32 %in) #1 {
 ; CHECK-NEXT:    ;;#ASMEND
+; CHECK-NEXT:    v_readlane_b32 s16, v22, 44
+; CHECK-NEXT:    v_readlane_b32 s17, v22, 45
+; CHECK-NEXT:    v_readlane_b32 s18, v22, 46
```
Hard to tell if this is a regression or just moved the spills around

arsenm commented on llvm/test/CodeGen/AMDGPU/spill-scavenge-offset.ll:
```diff
@@ -10112,6 +10094,24 @@ define amdgpu_kernel void @test_limited_sgpr(ptr addrspace(1) %out, ptr addrspac
 ; GFX6-NEXT:    v_readlane_b32 s7, v4, 7
 ; GFX6-NEXT:    buffer_load_dword v4, off, s[40:43], 0
 ; GFX6-NEXT:    s_waitcnt vmcnt(0)
+; GFX6-NEXT:    s_mov_b64 exec, s[24:25]
```
Regression

arsenm commented on llvm/lib/CodeGen/CalcSpillWeights.cpp:
```diff
@@ -188,6 +188,9 @@ float VirtRegAuxInfo::weightCalcHelper(LiveInterval &LI, SlotIndex *Start,
   bool ShouldUpdateLI = !IsLocalSplitArtifact;
 
+  // We will scale the weight by the register weight of register class.
+  unsigned Factor =
+      TRI.getRegClassWeight((MRI.getRegClass(LI.reg()))).RegWeight;
```
Can sink this down to the uses under the condition

arsenm commented on llvm/test/CodeGen/AMDGPU/hsa-metadata-kernel-code-props.ll:
```diff
@@ -47,10 +47,10 @@ entry:
-; GFX803:   .sgpr_spill_count: 10
-; GFX900:   .sgpr_spill_count: 62
-; GFX1010:  .sgpr_spill_count: 60
+; GFX700:   .sgpr_spill_count: 13
+; GFX803:   .sgpr_spill_count: 13
```
Regression on these targets

arsenm commented on llvm/test/CodeGen/AMDGPU/hsa-metadata-kernel-code-props.ll:
```diff
@@ -47,10 +47,10 @@ entry:
-; GFX1010:  .sgpr_spill_count: 60
+; GFX700:   .sgpr_spill_count: 13
+; GFX803:   .sgpr_spill_count: 13
+; GFX900:   .sgpr_spill_count: 57
+; GFX1010:  .sgpr_spill_count: 56
```
But improvement here

arsenm commented on llvm/test/CodeGen/Hexagon/regalloc-bad-undef.mir:
```diff
@@ -1,9 +1,9 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
```
Should pre-commit a switch to generated checks

lukel97 commented:
> I think the `frequency` and the `size` here are the same thing as long as we are using reg units to scale the weight. For example, for a LMUL 8 register, we can consider it as that all its 8 registers have one use, `frequency+=8` and also `size+=8`. 

Ah ok that makes sense to me at least for RISC-V, where a single LMUL 8 spill would be roughly the same as 8 * LMUL 1 spills. Does this hold for other targets though? E.g. could it be possible that spilling a register that uses two regunits costs the same as spilling another register that only uses one regunit?

wangpc-pp commented:
> > I think the `frequency` and the `size` here are the same thing as long as we are using reg units to scale the weight. For example, for a LMUL 8 register, we can consider it as that all its 8 registers have one use, `frequency+=8` and also `size+=8`.
> 
> Ah ok that makes sense to me at least for RISC-V, where a single LMUL 8 spill would be roughly the same as 8 * LMUL 1 spills. Does this hold for other targets though? E.g. could it be possible that spilling a register that uses two regunits costs the same as spilling another register that only uses one regunit?

Yeah, good point! I think X86 may be the one, it has a lot of overlapped registers. But from the perpective of spills/reloads bytes, it may still be profitable. cc @topperc WDYT?
@arsenm I think this can be an evidence that we shouldn't apply the scaling by weight behavior to all targets.

arsenm commented:
> Ah ok that makes sense to me at least for RISC-V, where a single LMUL 8 spill would be roughly the same as 8 * LMUL 1 spills. Does this hold for other targets though? E.g. could it be possible that spilling a register that uses two regunits costs the same as spilling another register that only uses one regunit?

It essentially does for AMDGPU. On most targets we end up splitting spills into 32-bit register (2-regunit, top one unaddressable) operations anyway. Newer targets can spill multiple registers at a time, but that's still N times more expensive

wangpc-pp commented:
Ping. I really want to push this forward. :-)

wangpc-pp commented:
Ping.

arsenm commented on llvm/lib/CodeGen/CalcSpillWeights.cpp:
```diff
@@ -335,6 +335,10 @@ float VirtRegAuxInfo::weightCalcHelper(LiveInterval &LI, SlotIndex *Start,
     TotalWeight *= 0.5F;
 
+  // Finally, we scale the weight by the number of register lanes.
+  unsigned Lanes = MRI.getRegClass(LI.reg())->getLaneMask().getNumLanes();
```
You're not supposed to directly use the lane mask values like this, you can only do overlap based tests

wangpc-pp replied:
I can change it back to use register class weight.

omern1 commented:
Hi, thanks for looking at this.

I appreciate the direction that you're taking with this but, as it currently stands, this patch is causing a regression in a very widely used game engine with the `x86_64-sie-ps5` target for us. I also looked through the X86 test diffs and there are a few regressions in there (though I didn't spot anything major). Broadly I think we should wait to enable this by default till it's been measured with a wider range of workloads on a wider range of targets. To that end, you can gate it behind a CL opt and make it opt-in for targets.

I'm also currently trying to get numbers on x86_64/linux, I'll post those here shortly.

I've got a few comments/questions about the patch though:
- I see that you've attempted many approaches in this PR, why have you settled on this one specifically (ie. scaling by register weight)?
- On X86 your current approach has very little impact on vector registers because their register weights aren't very different from the scalar registers. IMO this is quite a large drawback of the current approach.
- Since the patch has evolved after the benchmarks were done, IMO it'd be prudent to redo those benchmarks.

wangpc-pp commented:
> Hi, thanks for looking at this.
> 
> I appreciate the direction that you're taking with this but, as it currently stands, this patch is causing a regression in a very widely used game engine with the `x86_64-sie-ps5` target for us. I also skimmed through the X86 test diffs and there are a few regressions in there (though I didn't spot anything major). Broadly I think we should wait to enable this by default till it's been measured with a wider range of workloads on a wider range of targets. To that end, you can gate it behind a CL opt and make it opt-in for targets.

Yeah, this was exactly what I have done at first and I only enabled this for RISC-V. @arsenm suggested that this may be profitable for all targets and we should enable it by default. For current result/feedback we have received, I totally agree that we should make it a target hook and let the target decide the scaling factor because I find it is really hard to find a factor that is suitable for all targets. WDYT? @arsenm 

> 
> I'm also currently trying to get numbers on x86_64/linux, I'll post those here shortly.

Thank you very much! That will be very useful!

> 
> I've got a few comments/questions about the patch though:
> 
>     * I see that you've attempted many approaches in this PR, why have you settled on this one specifically (ie. scaling by register weight)?

As that I have mentioned above, I was trying different kinds of factor to find an universal one but it seems to be not that easy and so it fell back to the original one (scaling by register weight).
> 
>     * On X86 your current approach has very little impact on vector registers because their register weights aren't very different from the scalar registers. IMO this is quite a large drawback of the current approach. (Edit: Looking at the tablegen, the weights for all the vector registers on X86 are 1 so this approach has no impact on them).

I noticed that before, so I have tried to scale the weight by the number of register lanes and it was likely to be more suitable for X86. You may try that as well.

> 
>     * Since the patch has evolved after the benchmarks were done, IMO it'd be prudent to redo those benchmarks.

The previous result of benchmarks should still stand as the original PR was scaling the weight by register weight. I don't see any difference after these attempts and the final fallback.

omern1 commented:
SPEC 2017 on `x86_64-unknown-linux-gnu`:

Previous here refers to c6c647588f911770170d2f7975b325f3d70cf89b which your branch is currently based on. Current refers to a branch with your patch applied on top of c6c647588f911770170d2f7975b325f3d70cf89b.

Performance Regressions - execution_time | Δ | Previous | Current | σ | Δ  (B) | σ  (B)
-- | -- | -- | -- | -- | -- | --
SPEC/CINT2017speed/623.xalancbmk_s/623.xalancbmk_s | 2.52% | 41.593 | 42.640 | 0.268 | 0.00% | 0.268
SPEC/CINT2017rate/500.perlbench_r/500.perlbench_r | 1.58% | 18.305 | 18.594 | 0.047 | 0.00% | 0.047
SPEC/CINT2017rate/523.xalancbmk_r/523.xalancbmk_r | 1.25% | 41.897 | 42.419 | 0.195 | 0.00% | 0.195

Performance Improvements - execution_time | Δ | Previous | Current | σ | Δ  (B) | σ  (B)
-- | -- | -- | -- | -- | -- | --
SPEC/CINT2017speed/641.leela_s/641.leela_s | -3.62% | 63.991 | 61.673 | 0.389 | 0.00% | 0.389

Not really seeing a clear benefit.

> Yeah, this was exactly what I have done at first and I only enabled this for RISC-V. @arsenm suggested that this may be profitable for all targets and we should enable it by default. For current result/feedback we have received, I totally agree that we should make it a target hook and let the target decide the scaling factor because I find it is really hard to find a factor that is suitable for all targets. WDYT? @arsenm

Thanks, making it a target hook seems like a reasonable approach.

> The previous result of benchmarks should still stand as the original PR was scaling the weight by register weight. I don't see any difference after these attempts and the final fallback.

I missed that bit earlier, seems reasonable.

wangpc-pp commented:
Ping. @arsenm Can I make it a hook now?

arsenm commented:
> Ping. @arsenm Can I make it a hook now?

It's still not a hook, or that hook already exists. You can adjust the spill weights of the classes instead of adding another layer of control on top of the existing control

wangpc-pp commented:
> > Ping. @arsenm Can I make it a hook now?
> 
> It's still not a hook, or that hook already exists. You can adjust the spill weights of the classes instead of adding another layer of control on top of the existing control

No. I mean, yes we can adjust the spill weight of a register class, but we don't count it in when calculating the spill weights of LiveIntervals. And this PR is trying to do it.

preames commented:
Ok, I stumbled back into this by analyzing another case where LMUL8 spilling goes bad.  I'm going to briefly explain the issue I was chasing, and then return to the status of this patch.

If we have a case which uses masks intermixed with m8 code, but for some reason the mask can't always be placed in v0, we can end up with cases where the register allocator will fragment an m8, and then be unable to "fix" the mistake.  In the particular example, we happen to assign v8-v15, then v16-v23, then assign two masks to v24, and v25 respective.  Because one of the instructions uses a mask, v0 is unavailable for an m8 value.  Next we try to assign an m8, and find no registers are available.  The key thing here is that we should have evicted v24, and v25 placed the m8 in v24-v31, and then reassigned the two m1 values into v7, and v6 respectively.  The key problem which prevented the eviction is that the spill weight of the m8 was perceived to be lower than either of the m1s, much less both.  This is definitely backwards.  Applying this patch has exactly the desired effect, and undoes the fragmentation.  

In terms of understanding this patch, I find it helpful to think of the scaling as applying just after the call to LiveIntervals::getSpillWeight.  (We're multiply an add reduction, so this is  equivalent on real numbers.  On floats, it's "mostly so".)  What we're doing here is increasing the cost of the spill or fill instruction.  This is 100% accurate as a m8 spill does cost 8x a m1 spill.  

I have applied this patch locally, and manually reviewed the RISC-V changes.  All but a very small handful are improvements.  I want this change.  I also inspected the values returned by getRegClassWeight to confirm they were reasonable for the RISC-V register classes; they are.  

@wangpc-pp I think it's clear that we can't enable this for all targets at this point.  Given that, we need to move this back under a target hook.  You seemed a bit confused by @arsenm 's comments on that.  Do you wan to take a second attempt, or would you like me to put up an alternate patch?

preames commented on llvm/lib/CodeGen/CalcSpillWeights.cpp:
```diff
@@ -335,6 +335,11 @@ float VirtRegAuxInfo::weightCalcHelper(LiveInterval &LI, SlotIndex *Start,
     TotalWeight *= 0.5F;
 
+  // Finally, we scale the weight by the register weight of register class.
+  unsigned RegWeight =
```
Minor style point, do:
  const TargetRegisterClass *RC = MRI.getRegClass(LI.reg());
  TotalWeight *= TRI.getRegClassWeight(RC).RegWeight;

wangpc-pp commented:
> @wangpc-pp I think it's clear that we can't enable this for all targets at this point. Given that, we need to move this back under a target hook. You seemed a bit confused by @arsenm 's comments on that. Do you wan to take a second attempt, or would you like me to put up an alternate patch?

Big thanks for raising this up and the detailed analysis! I will make it a hook and retry it!

wangpc-pp committed:
[RegAlloc] Scale the spill weight by target factor

Currently, the spill weight is only determined by isDef/isUse and
block frequency. However, for registers with different register
classes, the costs of spilling them are different.

For example, for `LMUL>1` registers (in which, several physical
registers compound a bigger logical register), the costs are larger
than `LMUL=1` case (in which, there is only one physical register).

To solve this problem, a new target hook `getSpillWeightScaleFactor`
is added. Targets can override the default factor (which is `1.0`)
according to the register class.

For RISC-V, the factors are set to the `RegClassWeight` which is
used to track register pressure. The values of `RegClassWeight`
happen to be the number of register units.

I believe all of the targets with compounded registers can benefit
from this change, but only RISC-V is customized in this patch since
it has widely been agreed to do so. The other targets need more
performance data to go further.

Partially fixes #113489.

lukel97 reviewed (approved):
I added some statistics to collect the total "lmul" that is spilled and reloaded statically with this change:

```diff
diff --git a/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp b/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp
index 2fdf6bd36e88..f4b5a5b29e71 100644
--- a/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp
+++ b/llvm/lib/Target/RISCV/RISCVInstrInfo.cpp
@@ -18,6 +18,7 @@
 #include "RISCVSubtarget.h"
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Statistic.h"
 #include "llvm/Analysis/MemoryLocation.h"
 #include "llvm/Analysis/ValueTracking.h"
 #include "llvm/CodeGen/LiveIntervals.h"
@@ -43,6 +44,10 @@ using namespace llvm;
 #define GET_INSTRINFO_NAMED_OPS
 #include "RISCVGenInstrInfo.inc"
 
+#define DEBUG_TYPE "riscv-instr-info"
+STATISTIC(TotalLMULSpilled, "Total LMUL spilled");
+STATISTIC(TotalLMULReloaded, "Total LMUL reloaded");
+
 static cl::opt<bool> PreferWholeRegisterMove(
     "riscv-prefer-whole-register-move", cl::init(false), cl::Hidden,
     cl::desc("Prefer whole register move for vector registers."));
@@ -615,12 +620,16 @@ void RISCVInstrInfo::storeRegToStackSlot(MachineBasicBlock &MBB,
     IsScalableVector = false;
   } else if (RISCV::VRRegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VS1R_V;
+    TotalLMULSpilled += 1;
   } else if (RISCV::VRM2RegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VS2R_V;
+    TotalLMULSpilled += 2;
   } else if (RISCV::VRM4RegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VS4R_V;
+    TotalLMULSpilled += 4;
   } else if (RISCV::VRM8RegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VS8R_V;
+    TotalLMULSpilled += 8;
   } else if (RISCV::VRN2M1RegClass.hasSubClassEq(RC))
     Opcode = RISCV::PseudoVSPILL2_M1;
   else if (RISCV::VRN2M2RegClass.hasSubClassEq(RC))
@@ -706,12 +715,16 @@ void RISCVInstrInfo::loadRegFromStackSlot(
     IsScalableVector = false;
   } else if (RISCV::VRRegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VL1RE8_V;
+    TotalLMULReloaded += 1;
   } else if (RISCV::VRM2RegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VL2RE8_V;
+    TotalLMULReloaded += 2;
   } else if (RISCV::VRM4RegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VL4RE8_V;
+    TotalLMULReloaded += 4;
   } else if (RISCV::VRM8RegClass.hasSubClassEq(RC)) {
     Opcode = RISCV::VL8RE8_V;
+    TotalLMULReloaded += 8;
   } else if (RISCV::VRN2M1RegClass.hasSubClassEq(RC))
     Opcode = RISCV::PseudoVRELOAD2_M1;
   else if (RISCV::VRN2M2RegClass.hasSubClassEq(RC))
```

Running it against SPEC CPU 2017, -O3 -march=rva23u64, it looks like this helps a good bit on imagick/gcc/deepsjeng! 

```
Program                                       riscv-instr-info.TotalLMULReloaded                riscv-instr-info.TotalLMULSpilled               
                                              lhs                                rhs     diff   lhs                               rhs     diff  
FP2017rate/508.namd_r/508.namd_r                 6.00                               6.00   0.0%    1.00                              1.00   0.0%
INT2017spe...ed/620.omnetpp_s/620.omnetpp_s      5.00                               5.00   0.0%    4.00                              4.00   0.0%
INT2017spe...00.perlbench_s/600.perlbench_s      8.00                               8.00   0.0%    4.00                              4.00   0.0%
INT2017speed/625.x264_s/625.x264_s              43.00                              43.00   0.0%   47.00                             47.00   0.0%
INT2017rate/525.x264_r/525.x264_r               43.00                              43.00   0.0%   47.00                             47.00   0.0%
INT2017rat...23.xalancbmk_r/523.xalancbmk_r      6.00                               6.00   0.0%    6.00                              6.00   0.0%
INT2017rate/520.omnetpp_r/520.omnetpp_r          5.00                               5.00   0.0%    4.00                              4.00   0.0%
INT2017rat...00.perlbench_r/500.perlbench_r      8.00                               8.00   0.0%    4.00                              4.00   0.0%
FP2017speed/644.nab_s/644.nab_s                 25.00                              25.00   0.0%   25.00                             25.00   0.0%
FP2017speed/619.lbm_s/619.lbm_s                 42.00                              42.00   0.0%   42.00                             42.00   0.0%
FP2017rate/544.nab_r/544.nab_r                  25.00                              25.00   0.0%   25.00                             25.00   0.0%
FP2017rate/519.lbm_r/519.lbm_r                  42.00                              42.00   0.0%   42.00                             42.00   0.0%
FP2017rate/511.povray_r/511.povray_r           122.00                             122.00   0.0%   66.00                             66.00   0.0%
INT2017spe...23.xalancbmk_s/623.xalancbmk_s      6.00                               6.00   0.0%    6.00                              6.00   0.0%
FP2017speed/638.imagick_s/638.imagick_s       5054.00                            5053.00  -0.0% 4433.00                           3803.00 -14.2%
FP2017rate/538.imagick_r/538.imagick_r        5054.00                            5053.00  -0.0% 4433.00                           3803.00 -14.2%
FP2017rate/510.parest_r/510.parest_r          1349.00                            1343.00  -0.4% 1089.00                           1083.00  -0.6%
FP2017rate/526.blender_r/526.blender_r        1138.00                            1127.00  -1.0% 1064.00                           1025.00  -3.7%
INT2017spe...31.deepsjeng_s/631.deepsjeng_s    284.00                             274.00  -3.5%  154.00                            132.00 -14.3%
INT2017rat...31.deepsjeng_r/531.deepsjeng_r    284.00                             274.00  -3.5%  154.00                            132.00 -14.3%
INT2017speed/602.gcc_s/602.gcc_s               113.00                              83.00 -26.5%  107.00                             77.00 -28.0%
INT2017rate/502.gcc_r/502.gcc_r                113.00                              83.00 -26.5%  107.00                             77.00 -28.0%
INT2017rate/505.mcf_r/505.mcf_r                  0.00                               0.00                                                        
INT2017rate/541.leela_r/541.leela_r              0.00                               0.00                                                        
INT2017rate/557.xz_r/557.xz_r                    0.00                               0.00                                                        
INT2017speed/605.mcf_s/605.mcf_s                 0.00                               0.00                                                        
INT2017speed/641.leela_s/641.leela_s             0.00                               0.00                                                        
INT2017speed/657.xz_s/657.xz_s                   0.00                               0.00                                                        
                           Geomean difference                                             -3.1%                                            -5.8%
      riscv-instr-info.TotalLMULReloaded                         riscv-instr-info.TotalLMULSpilled                        
```

I think I was previously skeptical because I only measured the absolute number of spills/reloads, which included scalars + didn't show much of a change:

```
Program                                       regalloc.NumSpills                regalloc.NumReloads               
                                              lhs                rhs      diff  lhs                 rhs      diff 
FP2017rate/526.blender_r/526.blender_r        13411.00           13430.00  0.1% 27478.00            27509.00  0.1%
INT2017speed/602.gcc_s/602.gcc_s              11376.00           11381.00  0.0% 25795.00            25800.00  0.0%
INT2017rate/502.gcc_r/502.gcc_r               11376.00           11381.00  0.0% 25795.00            25800.00  0.0%
FP2017rate/508.namd_r/508.namd_r               6729.00            6729.00  0.0% 16370.00            16370.00  0.0%
FP2017rate/510.parest_r/510.parest_r          44293.00           44293.00  0.0% 87404.00            87404.00  0.0%
INT2017speed/641.leela_s/641.leela_s            310.00             310.00  0.0%   449.00              449.00  0.0%
INT2017speed/625.x264_s/625.x264_s             2147.00            2147.00  0.0%  4598.00             4598.00  0.0%
INT2017spe...23.xalancbmk_s/623.xalancbmk_s    1822.00            1822.00  0.0%  2969.00             2969.00  0.0%
INT2017spe...ed/620.omnetpp_s/620.omnetpp_s     719.00             719.00  0.0%  1210.00             1210.00  0.0%
INT2017speed/605.mcf_s/605.mcf_s                123.00             123.00  0.0%   372.00              372.00  0.0%
INT2017spe...00.perlbench_s/600.perlbench_s    4375.00            4375.00  0.0%  9740.00             9740.00  0.0%
INT2017rate/557.xz_r/557.xz_r                   300.00             300.00  0.0%   603.00              603.00  0.0%
INT2017rate/541.leela_r/541.leela_r             310.00             310.00  0.0%   449.00              449.00  0.0%
INT2017rate/525.x264_r/525.x264_r              2147.00            2147.00  0.0%  4598.00             4598.00  0.0%
INT2017rat...23.xalancbmk_r/523.xalancbmk_r    1822.00            1822.00  0.0%  2969.00             2969.00  0.0%
INT2017rate/520.omnetpp_r/520.omnetpp_r         719.00             719.00  0.0%  1210.00             1210.00  0.0%
INT2017rate/505.mcf_r/505.mcf_r                 123.00             123.00  0.0%   372.00              372.00  0.0%
INT2017rat...00.perlbench_r/500.perlbench_r    4375.00            4375.00  0.0%  9740.00             9740.00  0.0%
FP2017speed/644.nab_s/644.nab_s                 713.00             713.00  0.0%  1066.00             1066.00  0.0%
FP2017speed/619.lbm_s/619.lbm_s                  88.00              88.00  0.0%    90.00               90.00  0.0%
FP2017rate/544.nab_r/544.nab_r                  713.00             713.00  0.0%  1066.00             1066.00  0.0%
FP2017rate/519.lbm_r/519.lbm_r                   90.00              90.00  0.0%    92.00               92.00  0.0%
FP2017rate/511.povray_r/511.povray_r           1571.00            1571.00  0.0%  3043.00             3043.00  0.0%
INT2017speed/657.xz_s/657.xz_s                  300.00             300.00  0.0%   603.00              603.00  0.0%
FP2017speed/638.imagick_s/638.imagick_s        4074.00            4054.00 -0.5% 10335.00            10452.00  1.1%
FP2017rate/538.imagick_r/538.imagick_r         4074.00            4054.00 -0.5% 10335.00            10452.00  1.1%
INT2017rat...31.deepsjeng_r/531.deepsjeng_r     344.00             341.00 -0.9%   690.00              691.00  0.1%
INT2017spe...31.deepsjeng_s/631.deepsjeng_s     344.00             341.00 -0.9%   690.00              691.00  0.1%
                           Geomean difference                             -0.1%                               0.1%
```

So I guess that answers my worry that increasing LMUL 8 weights might increase smaller LMUL spills, i.e. it looks like it doesn't!

Thanks for being patient on this for so long, the code changes + test diff LGTM :)

preames reviewed (approved):
LGTM

(Make sure you rebase carefully when landing - there's enough test changes I suspect you'll easily get conflicts.)

arsenm commented:
I still think we should be making active effort to just make this the default. I believe this is another case where the target hook is just being used to paper over issues in other areas (I'm not objecting to this as an incremental step).

In particular in the previous revision of the patch, the AMDGPU behavior showed rematerialize was totally broken. I also think that something is wrong with whatever tablegen is doing to compute the default class weight. We have quite a lot of code scattered in the allocator and tablegen that were never written to account for subregisters. The defaults are wrong for the allocation priorities (which RISCV also does not look like it's trying to set). We also lack a splitting strategy to evict only a subregister when it would help, and spilling induces new liveness to dead lanes.

preames commented:
> I still think we should be making active effort to just make this the default. I believe this is another case where the target hook is just being used to paper over issues in other areas (I'm not objecting to this as an incremental step).

I agree with this, but FYI, your last clause had gotten lost in previous review comments, at least for me.  I do think it's important to highlight that the thing being papered over is for the target keeping the old behavior.  Long term, I do think all targets should have this particular behavior.  

> In particular in the previous revision of the patch, the AMDGPU behavior showed rematerialize was totally broken. I also think that something is wrong with whatever tablegen is doing to compute the default class weight. We have quite a lot of code scattered in the allocator and tablegen that were never written to account for subregisters. The defaults are wrong for the allocation priorities (which RISCV also does not look like it's trying to set). We also lack a splitting strategy to evict only a subregister when it would help, and spilling induces new liveness to dead lanes.

I've been looking into some of this, and trying to find ways to split things apart into some actionable pieces.  Incrementalism here is critical for keeping the problem vaguely understandable.  However, it sounds like you have a couple cases identified that I haven't yet stumbled into to.  Could we compare notes offline?

preames commented:
> The defaults are wrong for the allocation priorities (which RISCV also does not look like it's trying to set).

Thanks for catching this, I hadn't been aware of that mechanism.  Adding in https://github.com/llvm/llvm-project/pull/131176 though it doesn't seem to have too much effect.

arsenm commented:
> Could we compare notes offline?

Sure, I have a long list of regalloc issues (and each one takes forever to work through)
